# 01 – Data Quality Assessment



This notebook performs a comprehensive data quality audit of the NovaCred credit application dataset. As the Data Governance Task Force, our objective is to systematically identify, quantify, and remediate issues that may compromise the integrity, reliability, and fairness of downstream machine learning decisions.



### DATASET OVERVIEW

In [1]:
# We import Libraries 
import pandas as pd 
import os
import numpy as np
import json 

In [2]:
with open('../data/raw_credit_applications.json', 'r') as f:
    data = json.load(f) 

print(f"Total raw records: {len(data)}")
df = pd.DataFrame(data)


Total raw records: 502


In [3]:
# We can use json_normalize to flatten the nested JSON structure into a DataFrame
df = pd.json_normalize(data)
df.head(100)

,_id,spending_behavior,processing_timestamp,applicant_info.full_name,applicant_info.email,applicant_info.ssn,applicant_info.ip_address,applicant_info.gender,applicant_info.date_of_birth,applicant_info.zip_code,...,financials.credit_history_months,financials.debt_to_income,financials.savings_balance,decision.loan_approved,decision.rejection_reason,loan_purpose,decision.interest_rate,decision.approved_amount,financials.annual_salary,notes
0,app_200,"[{'category': 'Shopping', 'amount': 480}, {'ca...",2024-01-15T00:00:00Z,Jerry Smith,jerry.smith17@hotmail.com,596-64-4340,192.168.48.155,Male,2001-03-09,10036,...,23,0.20,31212,False,algorithm_risk_score,NaN,NaN,NaN,NaN,NaN
1,app_037,"[{'category': 'Rent', 'amount': 608}, {'catego...",NaN,Brandon Walker,brandon.walker2@yahoo.com,425-69-4784,10.1.102.112,M,1992-03-31,10032,...,51,0.18,17915,False,algorithm_risk_score,NaN,NaN,NaN,NaN,NaN
2,app_215,"[{'category': 'Rent', 'amount': 109}]",NaN,Scott Moore,scott.moore94@mail.com,370-78-5178,10.240.193.250,Male,1989-10-24,10075,...,41,0.21,37909,True,NaN,vacation,3.7,59000.0,NaN,NaN
3,app_024,"[{'category': 'Fitness', 'amount': 575}]",NaN,Thomas Lee,thomas.lee6@protonmail.com,194-35-1833,192.168.175.67,Male,1983-04-25,10077,...,70,0.35,0,True,NaN,NaN,4.3,34000.0,NaN,NaN
4,app_184,"[{'category': 'Entertainment', 'amount': 463}]",2024-01-15T00:00:00Z,Brian Rodriguez,brian.rodriguez86@aol.com,480-41-2475,172.29.125.105,M,1999-05-21,10080,...,14,0.23,31763,False,algorithm_risk_score,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,app_323,"[{'category': 'Travel', 'amount': 133}, {'cate...",NaN,Michael Thompson,michael.thompson45@gmail.com,862-33-2796,172.26.169.208,Male,08/11/1981,10090,...,92,0.40,27746,True,NaN,NaN,3.9,80000.0,NaN,NaN
96,app_288,"[{'category': 'Shopping', 'amount': 434}]",NaN,Susan Young,susan.young22@mail.com,159-19-1765,192.168.114.250,Female,1962-11-05,90252,...,71,0.27,23054,True,NaN,NaN,5.5,67000.0,NaN,NaN
97,app_396,"[{'category': 'Shopping', 'amount': 698}, {'ca...",NaN,Katherine Harris,katherine.harris16@outlook.com,174-24-9597,10.14.75.95,Female,1976-01-28,10027,...,124,0.20,33232,True,NaN,NaN,5.4,40000.0,NaN,NaN
98,app_371,"[{'category': 'Rent', 'amount': 180}, {'catego...",NaN,Sarah Walker,sarah.walker95@mail.com,142-97-2132,10.57.113.229,Female,1995-02-23,90250,...,12,0.11,59032,False,algorithm_risk_score,NaN,NaN,NaN,NaN,NaN


In [4]:
# We investigate summary statistics for numerical fields
df.describe()

,financials.credit_history_months,financials.debt_to_income,financials.savings_balance,decision.interest_rate,decision.approved_amount,financials.annual_salary
count,502.000000,502.000000,502.000000,292.000000,292.000000,5.000000
mean,50.402390,0.246195,29493.503984,4.564726,47845.890411,69200.000000
std,31.234824,0.136296,16775.309756,1.162866,18103.754530,22664.950915
min,-10.000000,0.050000,-5000.000000,2.500000,15000.000000,45000.000000
25%,27.250000,0.150000,17258.250000,3.500000,34000.000000,46000.000000
50%,48.000000,0.230000,27385.500000,4.550000,48000.000000,75000.000000
75%,72.000000,0.350000,38251.500000,5.600000,62250.000000,86000.000000
max,133.000000,1.850000,88078.000000,6.500000,80000.000000,94000.000000


In [5]:
# and a structured overview of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 502 entries, 0 to 501
Data columns (total 21 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   _id                               502 non-null    object 
 1   spending_behavior                 502 non-null    object 
 2   processing_timestamp              62 non-null     object 
 3   applicant_info.full_name          502 non-null    object 
 4   applicant_info.email              502 non-null    object 
 5   applicant_info.ssn                497 non-null    object 
 6   applicant_info.ip_address         497 non-null    object 
 7   applicant_info.gender             501 non-null    object 
 8   applicant_info.date_of_birth      501 non-null    object 
 9   applicant_info.zip_code           501 non-null    object 
 10  financials.annual_income          497 non-null    object 
 11  financials.credit_history_months  502 non-null    int64  
 12  financia

In [6]:
# We analyze the decision outcome distribution
approved = (df["decision.loan_approved"] == True).sum()
rejected = (df["decision.loan_approved"] == False).sum()
total = len(df)

print(f"Approved: {approved} ({approved/total*100:.1f}%)")
print(f"Rejected: {rejected} ({rejected/total*100:.1f}%)")
print(f"Total:    {total}")

Approved: 292 (58.2%)
Rejected: 210 (41.8%)
Total:    502


### *Overview Findings*
The dataset containts 502 credit application records that include applicant information, financial attributes, spending behavior categories and loan decision outcomes. 
The decision outcome is balanced - with 58.2% approved applications and 41.8% rejected. 

However some statitistics caught our eye:
- ` financials.annual_income` --> stored as object type rather than numeric - meaning that some string values are affecting data consistency 
- ` financials.credit_history_months`  --> the minimum value yielded -10 - implausible for credit history values 
- ` financials.savings_balance`  --> has a minimum of -5,000 - not common for a balance field 

### MISSINGNESS ANALYSIS

In [7]:
# Audit empty strings vs NaN before normalization
# Standard isna() does not detect empty strings or whitespace-only values

string_fields = ["applicant_info.email", "applicant_info.gender", "applicant_info.date_of_birth", 
                 "applicant_info.zip_code", "applicant_info.full_name", "applicant_info.ssn", "applicant_info.ip_address"]

empty_audit = pd.DataFrame({
    "nan_count": df[string_fields].isna().sum(),
    
    # Detect empty or whitespace-only strings safely
    "empty_string_count": df[string_fields]
        .apply(lambda col: col.astype(str).str.strip().eq("").sum())
}).assign(
    total_missing=lambda x: x["nan_count"] + x["empty_string_count"],
    total_missing_pct=lambda x: (x["total_missing"] / len(df) * 100).round(2)
)

# Show only affected fields
affected = empty_audit[empty_audit["empty_string_count"] > 0]

print("Fields containing empty or whitespace-only strings (not detected by isna()):")
display(affected)

Fields containing empty or whitespace-only strings (not detected by isna()):


,nan_count,empty_string_count,total_missing,total_missing_pct
applicant_info.email,0,7,7,1.39
applicant_info.gender,1,2,3,0.60
applicant_info.date_of_birth,1,4,5,1.00
applicant_info.zip_code,1,1,2,0.40


In [8]:
df[string_fields] = (
    df[string_fields]
        .replace(r'^\s*$', np.nan, regex=True)
)

In [9]:
# We check for missing values in each column + percentage of missing values per column
NaN_investigation = pd.DataFrame({'Missing_Count': df.isnull().sum(), 
                                  'Missing_Percent': (df.isnull().sum() / len(df) * 100).round(2)
                                  }).sort_values(by='Missing_Count', ascending=False)

display(NaN_investigation)

,Missing_Count,Missing_Percent
notes,500,99.60
financials.annual_salary,497,99.00
loan_purpose,452,90.04
processing_timestamp,440,87.65
decision.rejection_reason,292,58.17
decision.interest_rate,210,41.83
decision.approved_amount,210,41.83
applicant_info.email,7,1.39
applicant_info.ssn,5,1.00
financials.annual_income,5,1.00


### *Missingness Findings*

Many columns showed high missingness rates. The columns: ` Notes` ,`  Annual_salary` , ` Loan_purpose ` and ` Processing_timestamp`  have the highest rates.


# Remediations 
This section translates the findings into actions. It documents the key data quality issues identified during the audit of the NovaCred credit application dataset. Each issue is analyzed, quantified, and solved.

- [ 1 – Duplicate Records](#issue1)
- [ 2 – Column Inconsistency (Income vs Salary)](#issue2)
- [ 3 – Categorical Encoding (Gender)](#issue3)
- [ 4 – Time Column Inconsistencies](#issue4)
- [ 5 – Data Collection Failure (Invalid Values)](#issue5)

<a id="issue1"></a>

## Data Quality Issue 1 - Duplicate Records

In [10]:
# We investigate the duplicate records by ID (the primary key that needs to be unique for each application)
duplicate_count = df["_id"].duplicated().sum()
print(f"Duplicate ID count: {duplicate_count}")
display(df[df["_id"].duplicated(keep=False)].sort_values("_id"))

Duplicate ID count: 2


,_id,spending_behavior,processing_timestamp,applicant_info.full_name,applicant_info.email,applicant_info.ssn,applicant_info.ip_address,applicant_info.gender,applicant_info.date_of_birth,applicant_info.zip_code,...,financials.credit_history_months,financials.debt_to_income,financials.savings_balance,decision.loan_approved,decision.rejection_reason,loan_purpose,decision.interest_rate,decision.approved_amount,financials.annual_salary,notes
383,app_001,"[{'category': 'Fitness', 'amount': 576}]",NaN,Stephanie Nguyen,stephanie.nguyen47@mail.com,427-90-1892,10.121.120.213,Female,1986-05-27,90230,...,37,0.42,0,False,high_dti_ratio,NaN,NaN,NaN,NaN,NaN
455,app_001,"[{'category': 'Fitness', 'amount': 576}]",NaN,Stephanie Nguyen,stephanie.nguyen47@mail.com,NaN,NaN,NaN,NaN,NaN,...,37,0.42,0,False,high_dti_ratio,NaN,NaN,NaN,NaN,DUPLICATE_ENTRY_ERROR
8,app_042,"[{'category': 'Insurance', 'amount': 153}, {'c...",NaN,Joseph Lopez,joseph.lopez1@gmail.com,652-70-5530,192.168.91.142,Male,1990-05-04,10044,...,43,0.41,15974,False,algorithm_risk_score,NaN,NaN,NaN,NaN,NaN
354,app_042,"[{'category': 'Insurance', 'amount': 153}, {'c...",NaN,Joseph Lopez,joseph.lopez1@gmail.com,652-70-5530,192.168.91.142,Male,1990-05-04,10044,...,43,0.41,15974,False,algorithm_risk_score,NaN,NaN,NaN,NaN,RESUBMISSION


Two applications IDs are duplicated, in both cases the records are mostly identical with some exceptions.
This is a critical issue because the ` _id`  field is supposed to be a unique identifier for each record to ensure traceability and auditability.

In [11]:
# we found 2 duplicate records based on the "_id" column. 
# duplicates violate uniqueness constraints and compromise data integrity

duplicates = df[df["_id"].duplicated(keep=False)].sort_values("_id")

# we can see that some records have more missing values than others. We will keep the most complete record for each duplicate ID.
df["completeness_score"] = df.notnull().sum(axis=1)


df = df.sort_values(
    by=["_id", "completeness_score"],
    ascending=[True, False]
)

# we remove both the duplicate records and the completeness score column
df = df.drop_duplicates(subset="_id", keep="first")
df = df.drop(columns=["completeness_score"])

print("Remaining duplicate IDs:", df["_id"].duplicated().sum())

Remaining duplicate IDs: 0


<a id="issue2"></a>

## Data Quality Issue 2 – Column Inconsistency (annual_income vs annual_salary)


In [12]:
salary = df["financials.annual_salary"].notna()
income = df["financials.annual_income"].notna()

print(f"Records with annual_salary: {salary.sum()}")
print(f"Records with annual_income: {income.sum()}")
print(f"Records with both: {(salary & income).sum()}")
display(df.loc[salary, ["_id", "financials.annual_income", "financials.annual_salary"]])


Records with annual_salary: 5
Records with annual_income: 495
Records with both: 0


,_id,financials.annual_income,financials.annual_salary
94,app_421,NaN,46000.0
76,app_436,NaN,45000.0
149,app_449,NaN,75000.0
141,app_463,NaN,86000.0
99,app_479,NaN,94000.0


The 5 records missing ` annual_income` are exactly the 5 records where `annual_salary` is populated instead. No record has both fields.

This is a schema inconsistency: the same financial attribute was recorded under two different field names, likely due to a change in the data collection pipeline or inconsistent field mapping across intake systems.

In [13]:
df[df["financials.annual_salary"].notna()][
    ["financials.annual_income", "financials.annual_salary"]
]

# schema duplication or incomplete data pipeline integration - we fill income with salary only where income is missing (5 records
df["financials.income_unified"] = (
    df["financials.annual_income"]
    .combine_first(df["financials.annual_salary"])
)

# we check the number of records where income_unified is still missing
missing_income_unified = df["financials.income_unified"].isnull().sum()
print(f"Records with missing income_unified: {missing_income_unified}")

Records with missing income_unified: 0


<a id="issue3"></a>

## Data Quality Issue 3 – Categorical Encoding


In [14]:
# Analyze categorical encoding consistency for gender
gender_counts = (
    df["applicant_info.gender"]
    .fillna("NaN")
    .replace("", "Empty String")
    .value_counts(dropna=False)
    .reset_index()
)

gender_counts.columns = ["Gender Value", "Count"]

display(gender_counts)

,Gender Value,Count
0,Male,194
1,Female,193
2,F,58
3,M,53
4,NaN,2


From the displayed original dataframe it was noticeable the inconsistent encoding across `applicant_info.gender`records.

Categorical fields should have consistent encoding across all records. Variations complicate both aggregation and analysis. 

In [15]:
# we standardize the gender column to ensure consistent values M = male  
def standardize_gender(df):
    df["applicant_info.gender"] = (
        df["applicant_info.gender"]
        .astype(str)
        .str.strip()
        .str.lower()
    )
   
    df["applicant_info.gender"] = df["applicant_info.gender"].replace({
        "m": "male",
        "male ": "male",
        "f": "female",
        "female ": "female"
    })
    
    df.loc[df["applicant_info.gender"] == "nan", "applicant_info.gender"] = np.nan
    
    return df

df = standardize_gender(df)

# we check the value counts for gender to verify the standardization
df["applicant_info.gender"].value_counts()

applicant_info.gender
female    251
male      247
Name: count, dtype: int64

<a id="issue5"></a>

## Data Quality Issue 4 – Data Inconsistencies in Time Columns


In [16]:
# We check the unique values in the time columns to identify inconsistencies

# 440/500 records (88%) missing processing_timestamp
# we standardize the time format in processing_timestamp column to another format

df["processing_timestamp"] = pd.to_datetime(
    df["processing_timestamp"],
    errors="coerce",
    utc=True
)

total_records = len(df)

missing_ts = df["processing_timestamp"].isna().sum()

print(f"Missing timestamps: {missing_ts} ({missing_ts/total_records*100:.2f}%)")

Missing timestamps: 438 (87.60%)


In [17]:
# Analyze date format patterns in date_of_birth
import re

dob_values = df["applicant_info.date_of_birth"].dropna()

format_patterns = {
    "YYYY-MM-DD": r"^\d{4}-\d{2}-\d{2}$",
    "DD/MM/YYYY": r"^\d{2}/\d{2}/\d{4}$",
    "YYYY/MM/DD": r"^\d{4}/\d{2}/\d{2}$",
}

format_counts = {}
for label, pattern in format_patterns.items():
    format_counts[label] = dob_values.apply(lambda x: bool(re.match(pattern, str(x)))).sum()

unmatched = len(dob_values) - sum(format_counts.values())
format_counts["Other / Unmatched"] = unmatched

format_overview = pd.DataFrame(
    format_counts.items(),
    columns=["Date Format", "Count"]
)

format_overview["% of Non-Null"] = (format_overview["Count"] / len(dob_values) * 100).round(2)

print(f"Date format distribution for date_of_birth ({len(dob_values)} non-null values):")
display(format_overview)

Date format distribution for date_of_birth (496 non-null values):


,Date Format,Count,% of Non-Null
0,YYYY-MM-DD,339,68.35
1,DD/MM/YYYY,101,20.36
2,YYYY/MM/DD,56,11.29
3,Other / Unmatched,0,0.00


In [18]:
# Parsing
dob_series = df["applicant_info.date_of_birth"].astype(str).str.strip()

# First format
parsed_dob = pd.to_datetime(dob_series, format="%Y-%m-%d", errors="coerce")

# Second format
parsed_dob = parsed_dob.fillna(
    pd.to_datetime(dob_series, format="%d/%m/%Y", errors="coerce")
)

# Third format
parsed_dob = parsed_dob.fillna(
    pd.to_datetime(dob_series, format="%Y/%m/%d", errors="coerce")
)

invalid_dob = parsed_dob.isna().sum()

In this section, we audited temporal fields for completeness, consistency, and validity issues.  
The primary focus was on:

- `processing_timestamp`
- `applicant_info.date_of_birth`

Inconsistent date formats complicate parsing, comparison, and temporal analysis - standardization is required to avoid data ambiguity. 

<a id="issue5"></a>

## Data Quality Issue 5 – Data Collection Failure


In [19]:
# inspect records with negative credit_history_months or savings_balance
neg_credit = df[df["financials.credit_history_months"] < 0][["_id", "financials.credit_history_months", "financials.income_unified", "decision.loan_approved"]]
neg_savings = df[df["financials.savings_balance"] < 0][["_id", "financials.savings_balance", "financials.income_unified", "decision.loan_approved"]]

print(f"Records with negative credit_history_months:")
display(neg_credit)

print(f"\nRecords with negative savings_balance:")
display(neg_savings)

Records with negative credit_history_months:


,_id,financials.credit_history_months,financials.income_unified,decision.loan_approved
137,app_043,-10,131000,True
162,app_156,-3,25000,False



Records with negative savings_balance:


,_id,financials.savings_balance,financials.income_unified,decision.loan_approved
159,app_290,-5000,104000,True


Negative values in these fields represent **data validity violations**:

- Credit history length cannot be negative.
- Savings balance, as defined in this dataset, should not be negative 

Rather than imputing arbitrary values, we replaced these invalid entries with `NaN`.

These anomalies indicate either:
- Data entry errors  
- System validation failures during ingestion  
- Upstream transformation inconsistencies

In [20]:
# We set the illogical negative records we found to NaN 
df.loc[df["financials.credit_history_months"] < 0, "financials.credit_history_months"] = np.nan
df.loc[df["financials.savings_balance"] < 0, "financials.savings_balance"] = np.nan

print("Negative values after correction")
print(f"Negative credit_history_months: {(df['financials.credit_history_months'] < 0).sum()}")
print(f"Negative savings_balance: {(df['financials.savings_balance'] < 0).sum()}")

Negative values after correction
Negative credit_history_months: 0
Negative savings_balance: 0


## Conclusions 

The data quality assessment revealed structural weaknesses affecting completeness, consistency, validity, and auditability.

The most critical findings are:
1. **Duplicate primary keys** were identified, violating uniqueness constraints and compromising traceability.
2. A **schema inconsistency** split income across `annual_income` and `annual_salary`, indicating pipeline misalignment.
3. **Categorical encoding inconsistencies** in gender would have distorted fairness metrics if left unresolved.
4. **Severe timestamp missingness (87.6%)** weakens the audit trail of credit decisions.
5.  **negative financial values** logically impossible - indicates validation control gaps.



All issues were remediated through controlled standardization, consolidation, and integrity-preserving corrections. The dataset is now structurally consistent and suitable for downstream bias and governance analysis.

## Export Cleaned Dataset

In [21]:
# Export cleaned dataset for Bias and Governance analysis

# Final columns 

columns_to_drop = [
    "financials.annual_salary",
    "financials.annual_income",
    "notes",
    "processing_timestamp",
    "loan_purpose",
]

internal_cols = [c for c in df.columns if c.startswith("_")]

df_clean = df.drop(
    columns=[col for col in columns_to_drop if col in df.columns] + internal_cols,
    errors="ignore"
).reset_index(drop=True)

df_clean = df_clean.convert_dtypes()
for col in df_clean.columns:
    if "date" in col.lower() or "timestamp" in col.lower():
        df_clean[col] = pd.to_datetime(df_clean[col], errors="coerce", utc=True)

for col in df_clean.select_dtypes(include="object"):
    df_clean[col] = df_clean[col].astype("string")

    df_clean["applicant_info.date_of_birth"] = pd.to_datetime(
    df_clean["applicant_info.date_of_birth"],
    errors="coerce"
).dt.strftime("%Y-%m-%d")


output_path = "../data/cleaned_dataset.csv"

df_clean.to_csv(output_path, index=False)

print(f"Final cleaned dataset: {df_clean.shape[0]} rows x {df_clean.shape[1]} columns")
print(f"Exported to: {output_path}")

Final cleaned dataset: 500 rows x 16 columns
Exported to: ../data/cleaned_dataset.csv
